ProjFS.Mac/PrjFSKext/ProviderMessaging.cpp

```diff
@@ -0,0 +1,234 @@
+    // Mark all outstanding messages for this root as aborted and wake up the waiting threads
+    Mutex_Acquire(s_outstandingMessagesMutex);
+    {
+        OutstandingMessage* outstandingMessage;
+        LIST_FOREACH(outstandingMessage, &s_outstandingMessages, _list_privates)
+        {
+            if (outstandingMessage->rootHandle == providerVirtualizationRootHandle)
+            {
+                outstandingMessage->receivedResult = true;
+                outstandingMessage->result = MessageType_Result_Aborted;
+                wakeup(outstandingMessage);
+            }
+        }
+    }
+    Mutex_Release(s_outstandingMessagesMutex);
+}
+
+bool ProviderMessaging_TrySendRequestAndWaitForResponse(
+    VirtualizationRootHandle root,
+    MessageType messageType,
+    const vnode_t vnode,
+    const FsidInode& vnodeFsidInode,
+    const char* vnodePath,
+    int pid,
+    const char* procname,
+    int* kauthResult,
+    int* kauthError)
+{
+    // To be useful, the message needs to either provide an FSID/inode pair or a path
+    assert(vnodePath != nullptr || (vnodeFsidInode.fsid.val[0] != 0 || vnodeFsidInode.fsid.val[1] != 0));
+    bool result = false;
+    const char* relativePath = nullptr;
+    
+    OutstandingMessage message =
+    {
+        .receivedResult = false,
+        .rootHandle = root,
+    };
+    
+    if (nullptr != vnodePath)
+    {
+        relativePath = VirtualizationRoot_GetRootRelativePath(root, vnodePath);
+    }
+    
+    int nextMessageId = OSIncrementAtomic(&s_nextMessageId);
+    
+    Message messageSpec = {};
+    Message_Init(
+        &messageSpec,
+        &(message.request),
+        nextMessageId,
+        messageType,
+        vnodeFsidInode,
+        pid,
+        procname,
+        relativePath);
+
+    if (s_isShuttingDown)
+    {
+        return false;
+    }
+    
+    Mutex_Acquire(s_outstandingMessagesMutex);
+    {
+        LIST_INSERT_HEAD(&s_outstandingMessages, &message, _list_privates);
+    }
+    Mutex_Release(s_outstandingMessagesMutex);
+    
+    errno_t sendError = ActiveProvider_SendMessage(root, messageSpec);
+   
+    Mutex_Acquire(s_outstandingMessagesMutex);
+    {
+        if (0 != sendError)
+        {
+            // TODO: appropriately handle unresponsive providers
+            *kauthResult = KAUTH_RESULT_DEFER;
+        }
+        else
+        {
+            while (!message.receivedResult &&
+                   !s_isShuttingDown)
+            {
+                Mutex_Sleep(5, &message, &s_outstandingMessagesMutex);
+            }
+        
+            if (s_isShuttingDown)
+            {
+                *kauthResult = KAUTH_RESULT_DENY;
+            }
+            else if (MessageType_Response_Success == message.result)
+            {
+                *kauthResult = KAUTH_RESULT_DEFER;
+                result = true;
+            }
+            else
+            {
+                KextLog_Error("TrySendRequestAndWaitForResponse; message.result = %d", message.result);
```

Comment:
(This `KextLog_Error` looks new)

If my understanding is correct, this scenario is when the provider returns an error code from a callback?

If that's the case, I don't think we should be logging this as an error because it's a supported/expected scenario.

Reply:
Ah, good catch, looks like a local debugging commit snuck in there. I'll make sure to pull that back out.

Reply:
Just to confirm I've just done a diff of the old `KauthHandler.cpp` and the new `ProviderMessaging.cpp` and that was the only change that snuck into the moved functions apart from the function name changes.